pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,291 @@
+		e.sendPodLifecycleEvent(&PodLifecycleEvent{ID: types.UID(event.PodSandboxMetadata.Uid), Type: ContainerStarted, Data: event.ContainerId})
+		klog.V(4).InfoS("Received Container Started Event", "event", event)
+	case runtimeapi.ContainerEventType_CONTAINER_DELETED_EVENT:
+		// In case the pod is deleted it is safe to generate both ContainerDied and ContainerRemoved events, just like in the case of
+		// Generic PLEG. https://github.com/kubernetes/kubernetes/blob/24753aa8a4df8d10bfd6330e0f29186000c018be/pkg/kubelet/pleg/generic.go#L169
+		e.sendPodLifecycleEvent(&PodLifecycleEvent{ID: types.UID(event.PodSandboxMetadata.Uid), Type: ContainerDied, Data: event.ContainerId})
+		e.sendPodLifecycleEvent(&PodLifecycleEvent{ID: types.UID(event.PodSandboxMetadata.Uid), Type: ContainerRemoved, Data: event.ContainerId})
+		klog.V(4).InfoS("Received Container Deleted Event", "event", event)
+	}
+}
+
+func (e *EventedPLEG) getPodIPs(pid types.UID, status *kubecontainer.PodStatus) []string {
+	if len(status.IPs) != 0 {
+		return status.IPs
+	}
+
+	oldStatus, err := e.cache.Get(pid)
+	if err != nil || len(oldStatus.IPs) == 0 {
+		return nil
+	}
+
+	for _, sandboxStatus := range status.SandboxStatuses {
+		// If at least one sandbox is ready, then use this status update's pod IP
+		if sandboxStatus.State == runtimeapi.PodSandboxState_SANDBOX_READY {
+			return status.IPs
+		}
+	}
+
+	// For pods with no ready containers or sandboxes (like exited pods)
+	// use the old status' pod IP
+	return oldStatus.IPs
+}
+
+func (e *EventedPLEG) sendPodLifecycleEvent(event *PodLifecycleEvent) {
+	select {
+	case e.eventChannel <- event:
+	default:
+		// record how many events were discarded due to channel out of capacity
+		metrics.PLEGDiscardEvents.Inc()
+		klog.ErrorS(nil, "Evented PLEG: Event channel is full, discarded pod lifecycle event")
+	}
+}
+
+func getPodSandboxState(podStatus *kubecontainer.PodStatus) kubecontainer.State {
+	// increase running pod count when cache doesn't contain podID
+	var sandboxId string
+	for _, sandbox := range podStatus.SandboxStatuses {
+		sandboxId = sandbox.Id
+		// pod must contain only one sandbox
+		break
+	}
+
+	for _, containerStatus := range podStatus.ContainerStatuses {
+		if containerStatus.ID.ID == sandboxId {
+			if containerStatus.State == kubecontainer.ContainerStateRunning {
+				return containerStatus.State
+			}
+		}
+	}
+	return kubecontainer.ContainerStateExited
+}
+
+func (e *EventedPLEG) updateRunningPodMetric(podStatus *kubecontainer.PodStatus) {
+	cachedPodStatus, err := e.cache.Get(podStatus.ID)
+	if err != nil {
+		klog.ErrorS(err, "Evented PLEG: Get cache", "podID", podStatus.ID)
+	}
+	// cache miss condition: The pod status object will have empty state if missed in cache
+	if len(cachedPodStatus.SandboxStatuses) < 1 {
+		sandboxState := getPodSandboxState(podStatus)
+		if sandboxState == kubecontainer.ContainerStateRunning {
+			metrics.RunningPodCount.Inc()
```

Comment:
since the feature being on results in both plegs running, wont generic pleg and this conflict?  generic pleg does an explicit set for this value.

Reply:
see other comment, i think we need to have a pattern of list + watch for handling situations where we re-establish our lost watch to runtime (right now we only do the list 10 minutes later, but we should do it on every loss of watch)

Reply:
Both generic PLEG as well as evented PLEG makes use of the common pod cache (https://github.com/kubernetes/kubernetes/pull/111384/files#diff-67dd9e8f3ee257072765326cb4f242852554a2c0753563fa51e292c0a63a7b94R697-R700) which helps avoid the conflict. Generic pleg will reset the value of running pod count each time it performs the relist, so after fixed relist period we have the running pod count updated. In all other cases of evented pleg tacking action, while we increase/decrease this metric we rely upon event being received and subsequently checking for if the pod status as cached last time was already in running state or not.

Also, as @harche mentioned, now we have introduced re-list upon gRPC reconnection so each time there is a lost watch, we should be able to explicity set the pod count value from the whole pod list.